core/keyboard_nav/key_map.js

```diff
@@ -0,0 +1,250 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2019 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The namespace used to keep track of keyboard actions and the
+ * key codes used to execute those actions.
+ * This is used primarily for keyboard navigation.
+ */
+'use strict';
+
+goog.provide('Blockly.user.keyMap');
+
+/**
+ * Holds the serialized key to key action mapping.
+ * @type {Object<string, Blockly.KeyAction>}
+ */
+Blockly.user.keyMap.map_ = {};
+
+/**
+ * Holds the list of actions that can be executed using the keyboard.
+ * @type {Array<Blockly.KeyAction>}
+ */
+Blockly.user.keyMap.actions_ = [];
+
+/**
+ * Update the key map and actions array to contain the new action.
+ * @param {!string} keyCode The key code serialized by the serializeKeyEvent.
+ * @param {!Blockly.KeyAction} action The action to be executed when the keys
+ *     corresponding to the serialized key code is pressed.
+ * @package
+ */
+Blockly.user.keyMap.setActionForKey = function(keyCode, action) {
+  var key = Blockly.user.keyMap.getKeyByActionName(action.name);
+  // If the action already exists in the key map remove it and add the new mapping.
+  if (key) {
+    delete Blockly.user.keyMap.map_[key];
+  }
+  Blockly.user.keyMap.map_[keyCode] = action;
+  // If the action is not in the actions array add it.
+  if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+    Blockly.user.keyMap.actions_.push(action);
+  }
+};
+
+/**
+ * Creates a new key map and a new actions array.
+ * @param {Object<string, Blockly.KeyAction>} keyMap The object holding the key
+ *     to action mapping.
+ * @package
+ */
+Blockly.user.keyMap.setKeyMap = function(keyMap) {
+  var keys = Object.keys(keyMap);
+  // Update the actions array to include any new actions found in the new key map.
+  for (var i = 0; i < keys.length; i++) {
+    var action = keyMap[keys[i]];
+    if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+      Blockly.user.keyMap.actions_.push(action);
+    }
+  }
+  Blockly.user.keyMap.map_ = keyMap;
+};
+
+/**
+ * Gets the current key map.
+ * @return {Object<string,Blockly.KeyAction>} The object holding the key to
+ *     action mapping.
+ * @package
+ */
+Blockly.user.keyMap.getKeyMap = function() {
+  return Blockly.user.keyMap.map_;
```

Comment:
Should this return a copy of the map to make it clear you shouldn't edit it directly?